DataBiosphere/terra-cli
# [BENCH-547] Re-write AWS Config File on Resource Add

STATUS

- [done] implementation changes, manual testing
- [done] unit test updates
- Note: changes to flieUtils function needed for config file deletion. 

-----
CHANGES - 

manual configure:
- no change in command and user facing bahavior
- Move implementation from command.execute to AwsConfiguration file 
- Store the config details (such as path, cache etc) as comments in the config file

workspace create:
- Create a empty config file (no resources) with default config values

workspace delete:
- Delete the config file

resource create:
- Load the config from disk, add resource and write back the config file
- No changes to default resource config, since this is automatically added and default must be manually set

resource delete:
- Load the config from disk, delete resource and write back the config file
- If resource was default, remove the default resource config

-----

MANUAL TESTING - 
```
// create workspace id: 0026e5db-f44e-40c6-898f-43869842d766 / userFacingId: dex-ws-0828-1q
>> cat /Users/dexamundsen/.terra/aws/0026e5db-f44e-40c6-898f-43869842d766.conf 
#terraPath=/usr/local/bin/terra, awsVaultPath=/usr/local/bin/aws-vault, cacheWithAwsVault=false

// create resource
>> terra resource create s3-storage-folder --name=dex-s3-0828-2
>> cat /Users/dexamundsen/.terra/aws/0026e5db-f44e-40c6-898f-43869842d766.conf 
#terraPath=/usr/local/bin/terra, awsVaultPath=/usr/local/bin/aws-vault, cacheWithAwsVault=false
[profile dex-s3-0828-2] ..
[profile dex-s3-0828-2-ro] ..


// manual configure
>> terra workspace configure-aws --default-resource=dex-s3-0828-2 --cache-with-aws-vault=true
>> cat /Users/dexamundsen/.terra/aws/0026e5db-f44e-40c6-898f-43869842d766.conf 
#terraPath=/usr/local/bin/terra, awsVaultPath=/usr/local/bin/aws-vault, cacheWithAwsVault=true, defaultResourceName=dex-s3-0828-2
[profile dex-s3-0828-2] ..
[profile dex-s3-0828-2-ro] ..
[profile dex-s3-0828-2_] ..
[profile dex-s3-0828-2-ro_] ..
[default]
region = us-east-1
credential_process = /usr/local/bin/aws-vault export --no-session --format json dex-s3-0828-2-ro_

// create another resource
>> cat /Users/dexamundsen/.terra/aws/0026e5db-f44e-40c6-898f-43869842d766.conf 
.. contains config for new resource as expected (including caching profiles)
.. default is unchanged as expected

// delete default resource
>> terra resource delete --name=dex-s3-0828-2 
>> cat /Users/dexamundsen/.terra/aws/0026e5db-f44e-40c6-898f-43869842d766.conf 
#terraPath=/usr/local/bin/terra, awsVaultPath=/usr/local/bin/aws-vault, cacheWithAwsVault=true
.. entries for resource removed
.. config does not contain default resource as expected

// delete all resources
>> cat /Users/dexamundsen/.terra/aws/0026e5db-f44e-40c6-898f-43869842d766.conf 
#terraPath=/usr/local/bin/terra, awsVaultPath=/usr/local/bin/aws-vault, cacheWithAwsVault=true

// delete workspace
>> terra workspace delete
>> cat /Users/dexamundsen/.terra/aws/0026e5db-f44e-40c6-898f-43869842d766.conf
.. file not found as expected
```

![image](https://github.com/DataBiosphere/terra-cli/assets/2914285/97a35c9d-35c6-480c-9071-2ceee4ec9776)
